### PR TITLE
Feature/prefix globals warning

### PR DESCRIPTION
**Pull request by dingo-d**
Since we've released the Theme Sniffer plugin, more and more authors and reviewers are complaining about the prefix all globals sniff. 

This PR downgrades it to a warning instead of an error.

### CODE REVIEW

**Comment by dingo-d** [comment]
I just saw your comment @jrfnl in #189.

I think that rewording of the error message could help, but we would still get a lot of false positives.

**Comment by jrfnl** [comment]
> I think that rewording of the error message could help, but we would still get a lot of false positives.

Correct me if I'm wrong, but from what I've seen from the discussion, you're not getting false positives, but you want to allow non-prefixed variables in a certain context.

You should still want to forbid non-prefixed classes, global functions etc.

If that's the case, there are two options:
1. Either downgrade a specific _errorcode_ to a warning, not the whole sniff.
2. Or (better) extend the sniff and overload the relevant method to bow out early in case of theme template file, but still trigger an error in a other cases.

The fact that people are getting annoyed with the message, is not a fault of the sniff, but a case of educating people to write better code.

**Comment by dingo-d** [comment]
Yeah, the issue (from what I've seen) is that non prefixed variables in the template files trigger this sniff. And since there is no required way of loading the templates (like having them all in a specific folder in a theme called `templates`), it's not possible for the sniffer to know what is a template file and what isn't.

I'll try to get a hold of a few themes and see where the faulty code is, and then try to implement your suggestions. I'll work in this PR so it can stay open.

Thanks for the suggestions and advice :+1:

**Comment by jrfnl** [comment]
> it's not possible for the sniffer to know what is a template file and what isn't

I had a little think about that and I think we can probably detect whether something is a template file with a reasonably high degree of certainty (providing the customized sniff is only run on themes).

I'm thinking along the following lines:
* File isn't called `functions.php`.
* File contains inline HTML.
* File does _not_ contain declarations for classes, functions and such.

Does that help ?

**Comment by DannyCooper** [comment]
I might be underthinking this, but wouldn't it be possible to have a list of 'template-files' that would be right 90%+ of the time?

```
/template-parts/*
/templates/*
taxonomy-*
category-*
tag-*
page-*
archive-*
single-*
sidebar-*
header.php
footer.php
index.php
comments.php
search.php
searchform.php
404.php
```

Sure there would be edge cases, but if it eliminated 90% of the 'false-positives', would it be worth it?

**Comment by joyously** [comment]
I don't think a whitelist is a good idea. I agree with detecting that it's a template. In fact, my theme would not pass because my template files are output as widgets, so would not match the whitelist at all.
And the other half of the problem is that those template files still should not be modifying WordPress globals, so they can't just be skipped. _(or is that two different sniffs?)_

**Comment by dingo-d** [comment]
I think that the globals sniff would just have specific error codes downgraded to warnings, but a separate sniff would be used for the templates and checking if globals are present there.

**Comment by jrfnl** [comment]
> wouldn't it be possible to have a list of 'template-files' that would be right 90%+ of the time?

> I don't think a whitelist is a good idea. I agree with detecting that it's a template. In fact, my theme would not pass because my template files are output as widgets, so would not match the whitelist at all.

@DannyCooper @joyously Both approaches could be combined, a file whitelist first, for files not on that list: check the contents of the file.
If I remember correctly, the WPCS Filename sniff already has a template file name regex which could possibly be used or if not used directly, used a inspiration.

> And the other half of the problem is that those template files still should not be modifying WordPress globals, so they can't just be skipped. (or is that two different sniffs?)

Those are two different sniffs, so that shouldn't be a problem and the above would only be applied to the one sniff, the files would not be skipped for any other sniffs.

**Comment by dingo-d** [comment]
@jrfnl I'd like to work on this so that I can brush up on writing the sniffs, but I'd need some help 🙂 

You've mentioned downgrading certain errorcode to a warning. That would be done in `phpcs.xml.dist` or?

About sniff extension, I've seen `WordPress.WP.GlobalVariablesOverride.OverrideProhibited` triggered in template files, so the steps to override this would be:

1. Create a new sniff, say `GlobalVariablesOverrideTemplatesExclude` that extends the `GlobalVariablesOverrideSniff`

```php
use WordPressCS\WordPress\Sniffs\WP\GlobalVariablesOverrideSniff;

class GlobalVariablesOverrideTemplatesExclude extends GlobalVariablesOverrideSniff {
 /* [...] */
}
```

2. In the sniff add two regexes for template files and folders check. I took the one from `FileNameSniff` and modified it slightly (based on the comment by Danny Cooper)

```php
const THEME_EXCEPTIONS_REGEX = '`
		^                    # Anchor to the beginning of the string.
		(?:
							 # Template prefixes which can have exceptions.
			(?:archive|category|content|embed|page|single|tag|taxonomy|template|tag|sidebar|front-page|header|footer|index|comments|search|searchform|404)
			-[^\.]+          # These need to be followed by a dash and some chars.
		)\.(?:php|inc)$      # End in .php (or .inc for the test files) and anchor to the end of the string.
	`Dx';
```

and for templates (could probably be improved, I'm not a regex expert, but when I test this with regex101 it seems like it makes too much passes)

```php
const THEME_TEMPLATE_FOLDERS_REGEX = '`(?:template-parts/|templates/|partials/)`Dx';
```

3. Copy the entire `process_token()` method, but add two regex checks for the file path and file name at the beginning

```php
$file = $this->strip_quotes( $this->phpcsFile->getFileName() ); // File path that can contain one of the folders from regex.
$fileName = basename( $file ); // File name to check with the first regex.

// If the regex matches skip checking for globals override.
if ( preg_match( self::THEME_EXCEPTIONS_REGEX, $fileName ) === 1 ) {
  return;
}

if ( preg_match( self::THEME_TEMPLATE_FOLDERS_REGEX, $file ) === 1 ) {
  return;
}
```

Am I on the right track?

**Comment by dingo-d** [comment]
@jrfnl Just pinging to check up on this 🙂

**Comment by jrfnl** [comment]
> Am I on the right track?

@dingo-d Sorry, but no, not really. Might be easier if we discuss this over Slack ? Ping me if you like with a day/time.

**Comment by dingo-d** [comment]
I've updated the code after our call @jrfnl 

The current sniff overloads the `PrefixAllGlobals` sniff from `WPCS`. It looks if the file path contains certain folder names (`template-parts`, `templates` and `partials`). The folders array is set as public property so it can be extended if needed.

Also, the file name is checked against the regex in the `FileNameSniff`.

The tests are not passing so I'd need some help to see if I'm on the right track here, and what am I doing wrong.

And I've added 

```xml
<!-- Exclude PrefixAllGlobals Sniff from WPCS and use the one from this ruleset -->
<rule ref="WordPress.NamingConventions.PrefixAllGlobals">
  <severity>0</severity>
</rule>
```

in the `ruleset.xml`

Because 

```xml
<rule ref="WordPress">
  <exclude name="WordPress.NamingConventions.PrefixAllGlobals"/>
</rule>
```
Would load all sniffs from the `WordPress` standard, and we don't want that.

I'm not 100% sure this is ok, and if this will silence errors when we don't want them to be silenced in the case of `WordPress.NamingConventions.PrefixAllGlobals` sniff.

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ This projects adheres to [Semantic Versioning](https://semver.org/) and [Keep a
 
 _No documentation available about unreleased changes as of yet._
 
+## [0.2.0] - 2019-04-13
```
Comment (jrfnl) [inline CHANGELOG.md]: This shouldn't be part of this PR - other changes have happened since as well, so it makes more sense to add the changelog separately for all changes in one go.

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	 */
+	private $expected_results = array(
+
+		/*
+		 * In /CoreFunctionality/ThemeTemplatesException.
+		 */
+
+		// Files in a theme context.
+		'footer_widgets.inc' => 1,
+		'header.inc'         => 1,
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: You should probably also add some test files using the `allowed_folders` without a typical template name, i.e. `/partials/something.php`.

And a test file which _should_ still trigger the error, i.e. a file with a non-typical template name, like `something.php`.

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+ * @since 0.2.0
+ */
+class PrefixAllGlobalsUnitTest extends AbstractSniffUnitTest {
+
+	/**
+	 * Error files with the expected nr of errors.
+	 *
+	 * @var array
+	 */
+	private $expected_results = array(
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: This array can be dropped.

Instead of testing that we are receiving errors, we are testing that errors will _not_ be thrown.

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+	/**
+	 * Returns the lines where errors should occur.
+	 *
+	 * @param string $testFile The name of the file being tested.
+	 * @return array <int line number> => <int number of errors>
+	 */
+	public function getErrorList( $testFile = '' ) {
+
+		if ( isset( $this->expected_results[ $testFile ] ) ) {
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: See my remarks above. Only the test for the non-template file should trigger an error, none of the others should.

File: WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<!-- Annotation has to be on the second line as errors are thrown on line 1 and errors on annotation lines are ignored. -->
+phpcs:set WordPress.Files.FileName is_theme true
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc]: Property is for the wrong sniff.

Try:
```
// phpcs:set WPThemeReview.CoreFunctionality.PrefixAllGlobals prefixes[] my_theme
```

File: WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+<?php
+/* phpcs:set WordPress.Files.FileName is_theme false */
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc]: And to clear it:
```
// phpcs:set WPThemeReview.CoreFunctionality.PrefixAllGlobals prefixes[]
```

File: WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc [hunk 7]
```diff
@@ -0,0 +1,4 @@
+<!-- Annotation has to be on the second line as errors are thrown on line 1 and errors on annotation lines are ignored. -->
+phpcs:set WordPress.Files.FileName is_theme true
+<?php
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc]: You need to add some code to trigger the sniff so it will actually be tested.
Try something like:
```php
$my_theme_var = 123; // OK, prefixed.
$var = // OK, template file.
```

File: WPThemeReview/ruleset.xml [hunk 8]
```diff
@@ -150,4 +150,10 @@
 	<!-- Themes should never touch the timezone. -->
 	<rule ref="WordPress.WP.TimezoneChange"/>
 
+	<!-- Exclude PrefixAllGlobals Sniff from WPCS and use the one from this ruleset -->
```
[the remark this answers is not in the document]
Comment (dingo-d) [reply WPThemeReview/ruleset.xml]: Makes much more sense. I wasn't sure if I need this because I want to have its functionality, but it will be included via the new sniff (only slightly changed for themes). :+1:

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+ * @since 0.2.0
+ */
+class PrefixAllGlobalsSniff extends WPCSPrefixAllGlobalsSniff {
+
+	/**
+	 * List of allowed folders to check the file path against
+	 *
+	 * @var array
+	 */
+	public $allowed_folders = [
```
[the remark this answers is not in the document]
Comment (dingo-d) [reply WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: In fact it was. If I want to create my own ruleset based on this rule, I can maybe add some additional folders (not all people may name the template folders that way). 

That was the rationale behind making it public 🙂 I can change it to `protected` if you think that the better way would be, when working on your own standard, to extend this sniff and just change this property.

If non of those make sense than I can make it `private` and if we see some extra folder names popping up in the reviews, we can easily add those in the array.

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	/**
+	 * Checks if the given file is located in the $allowed_folders array
+	 *
+	 * @param  string $file Full path of the sniffed file.
+	 * @return boolean
+	 */
+	private function is_from_allowed_folder( $file ) {
+		$result = array_filter(
```
[the remark this answers is not in the document]
Comment (dingo-d) [reply WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: Thanks for the clarification. I'll replace this with the suggestion. 

Would it change anything if I passed by reference to the file path in the `use` statement? 

I thought it should work. because I'm calling this method on every file, so the `$file` parameter changes every time the sniff is ran. Although I'm probably thinking about the process of sniffing in the wrong way - I'm not redefining the whole thing for each file, just once at the beginning of the sniff process, right? That's why my way didn't work. Still, even if I used pass-by-refference it would still be wrong because of the third point.

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+		}
+	}
+
+	/**
+	 * Checks if the given file is located in the $allowed_folders array
+	 *
+	 * @param  string $file Full path of the sniffed file.
+	 * @return boolean
+	 */
+	private function is_from_allowed_folder( $file ) {
```
[the remark this answers is not in the document]
Comment (dingo-d) [reply WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: I'm reading a lot of _Clean Code_ by Uncle Bob Martin, so it kinda became a habit to split parts of the business logic to separate private methods 😄 . Looks a bit cleaner, plus I can then use the private method in the one conditional in the `process_variable_assignment()` method.

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+		if ( 'STDIN' === $file ) {
+			return;
+		}
+
+		$fileName = basename( $file );
+
+		// Don't process in case of a template file or folder.
+		if ( preg_match( FileNameSniff::THEME_EXCEPTIONS_REGEX, $fileName ) !== 1 && $this->is_from_allowed_folder( $file ) ) {
+			return;
+		} else {
```
[the remark this answers is not in the document]
Comment (dingo-d) [reply WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: My bad, totally overlooked this part 😬

File: CHANGELOG.md (hunk 1 above)
Comment (dingo-d) [reply CHANGELOG.md]: Will remove this definitely

File: WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc (hunk 7 above)
Comment (dingo-d) [reply WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc]: Although this should be in the `header.inc` as that is considered a template file, while `footer_widgets.inc` isn't, right?

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php (hunk 4 above)
Comment (dingo-d) [reply WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: So I should remove this. But how do I test for the specific `.inc` files if they are triggering errors? Like `footer_widgets.inc` and say `social-share.inc` (outside of `partials` folder) that should also trigger an error.

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: It's fine to leave it as `public` if that's the intended behaviour.

In that case, you may want to adjust the property description to make that clear.
We may also need to have a think about whether the default value should in that case be an empty array instead ?

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: > Would it change anything if I passed by reference to the file path in the use statement?

Yes, it would, but point 1 and 3 would still stand, so it still wouldn't work.

> I'm not redefining the whole thing for each file, just once at the beginning of the sniff process, right?

Correct.

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php (hunk 11 above)
[the remark this answers is not in the document]
Comment (jrfnl) [reply WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: That's fine, I was just wondering.

Typical reasons for splitting things off to functions are (and most of these don't apply):
1. Removed code duplication - but the code in this case is only being used once.
2. Make it easier to write unit tests for a particular part of the code - but we do integration testing for sniffs, not unit testing.
3. Code readability.

File: WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc (hunk 7 above)
Comment (jrfnl) [reply WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc]: It should be in all the test files for this sniff, but the `OK` should be changed to `Bad` for non-template files.

If the code is not in non-template files, the sniff wouldn't be triggered and we still wouldn't be testing if the "fall through to the parent sniff if the file is not a template file" works correctly.

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php (hunk 4 above)
Comment (jrfnl) [reply WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: The WPCS file name sniff is the only one to include a list like that in a property as all the errors would be on the same line and it improved the documentation of why and what.

Normally for sniffs which need to be tested using several files, the test method would use a unit test file name based `switch` which also allows more easily to the differentiate between the lines on which errors/warnings are being thrown between files.
The `default` case would be an empty array (explicitly expect no errors), so you would only need to list the non-template test case files in the `switch`.

Like in the file you linked above as well as in, for instance, this one: https://github.com/WPTRT/WPThemeReview/blob/develop/WPThemeReview/Tests/ThouShallNotUse/NoAutoGenerateUnitTest.php

> Should I manually include the test files,

'The `switch` doesn't include the test files. Test case files with a `SniffNameTest.1.inc` (numbered) file name will be included for testing automatically by PHPCS, it's only for non-standard test case file names (which you need for this sniff) that you have to tell PHPCS which files to use in the test and you already do so via the `getTestFiles()` method.

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (dingo-d) [reply WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: Would something like this be ok?

```php
/**
 * The list of allowed folders to check the file path against.
 *
 * This array can be extended in the custom ruleset.
 * If you are using PHPCS prior to the version 3.4.0 you would overwrite the current value by setting:
 *
 * <property name="allowed_folders" type="array" values="template-parts,templates,partials,my-template-folder">
 *
 * If you are using PHPCS 3.4.0 or greater you can extend this array by setting extends="true" attribute:
 *
 * <property name="allowed_folders" type="array" extends="true" values="my-template-folder">
 *
 * @var array
 */
```

If we set empty array as a default value we would define the folders inside the `ruleset.xml`? I'm  for it 🙂

**Comment by dingo-d** [comment]
One thing that I kinda realised now is that `THEME_EXCEPTIONS_REGEX` from the `FileNameSniff` doesn't check for all the cases mentioned in the [comment by Danny Cooper](https://github.com/WPTRT/WPThemeReview/pull/208#issuecomment-478541659).

Because of that the `header.inc` file is flagged as a non template file and I get an error in the tests.

I remember you mentioning that this change should be done upstream, but we'd have to wait for this, and for the new version of WPCS, and I'd like to fix this asap so that Theme Sniffer development can continue (we have quite some work to be done there).

Would it be better to define the regex in the sniff here, and once this is fixed upstream, we should just replace it with `FileNameSniff::THEME_EXCEPTIONS_REGEX`?

File: WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	/**
+	 * Checks if the given file path is located in the $allowed_folders array
+	 *
+	 * @param  string $path Full path of the sniffed file.
+	 * @return boolean
+	 */
+	private function is_from_allowed_folder( $path ) {
+		foreach ( $this->allowed_folders as $folder ) {
```
Comment (jrfnl) [inline WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php]: If the default for `$allowed_folders` is going to be an empty array, I'd suggest adding the below code at the top of the function.
```php
if ( empty( $this->allowed_folders ) || ! is_array( $this->allowed_folders ) ) {
    return false

}
```

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+	 * @return array <int line number> => <int number of errors>
+	 */
+	public function getErrorList( $testFile = 'partials/post-edit.inc' ) {
+		switch ( $testFile ) {
+			case 'social-share.inc':
+				return array(
+					6 => 1,
+				);
+			case 'header.inc':
+				// Template file - all OK, fall through to the default case.
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: This comment seems incorrect.

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+			case 'footer_widgets.inc':
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: No need for duplicate code here if the line numbers in the test files are the same everywhere.
```php
			case 'social-share.inc':
			case 'header.inc':
			case 'footer_widgets.inc':
				return array(
					6 => 1,
				);
```

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+					6 => 1,
+				);
+			case 'post-edit.inc':
+				// Template file - all OK, fall through to the default case.
+			default:
+				return array();
+		}
+	}
+
+	/**
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: As a convention, this method should be above the `getErrorList()`/`getWarningList()` methods (as it will be called before either of them).

File: WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/header.inc [hunk 17]
```diff
@@ -0,0 +1,7 @@
+<!-- Annotation has to be on the second line as errors are thrown on line 1 and errors on annotation lines are ignored. -->
+// phpcs:set WPThemeReview.CoreFunctionality.PrefixAllGlobals prefixes[] my_theme
+<?php
+
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/header.inc]: Why is the `$my_theme_var = 123; // OK, prefixed.` snippet missing here ?

File: WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc [hunk 18]
```diff
@@ -0,0 +1,7 @@
+<!-- Annotation has to be on the second line as errors are thrown on line 1 and errors on annotation lines are ignored. -->
```
Comment (jrfnl) [inline WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/footer_widgets.inc]: The above comment can be removed from all test files for this sniff. For this sniff, it's fine to have the annotation on line 1.

The "_errors are thrown on line 1_" comes from the FileName sniff, where that is true, but that is not the case for this sniff (errors are thrown on the line where the variable is declared).

File: composer.json [hunk 19]
```diff
@@ -38,7 +38,7 @@
 	],
 	"require"    : {
 		"php"                      : ">=5.4",
-		"squizlabs/php_codesniffer": "^3.3.1",
+		"squizlabs/php_codesniffer": "^3.4.1",
```
Comment (jrfnl) [inline composer.json]: Any particular reason to change the minimum requirement ? And if so, why to this particular version ?

File: WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php (hunk 16 above)
Comment (dingo-d) [reply WPThemeReview/Tests/CoreFunctionality/PrefixAllGlobalsUnitTest.php]: Yeah, this makes more sense 🙂

File: WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/header.inc (hunk 17 above)
Comment (dingo-d) [reply WPThemeReview/Tests/CoreFunctionality/ThemeTemplatesException/header.inc]: Forgot to add it, my bad.

File: composer.json (hunk 19 above)
Comment (dingo-d) [reply composer.json]: I think this was the latest version when I started to work on it initially. I can up this to `^3.4.2`.

**Comment by dingo-d** [comment]
I've fixed the issues, the tests should be passing now. If all is ok (I really hope it is 😂), you can squash commits and merge them to avoid the commit clutter 😄

File: composer.json (hunk 19 above)
Comment (jrfnl) [reply composer.json]: But why change the minimum version at all at this time ? Is there anything in the newer PHPCS which this PR needs ? which TRTCS needs ? If not,why not keep it as, in line with the WPCS minimum requirements ?

**Comment by jrfnl** [comment]
> I've fixed the issues, the tests should be passing now. If all is ok (I really hope it is)

Nearly there. Just the question about the PHPCS dependency change + the defaults for the `$allowed_folders` still being in the sniff remaining.

File: composer.json (hunk 19 above)
Comment (dingo-d) [reply composer.json]: Ok, makes sense 🙂It's just a habit of mine to keep the libraries up to date (watching for BC breaks ofc).

**Comment by dingo-d** [comment]
I think this should be it 🙂

File: composer.json (hunk 19 above)
Comment (jrfnl) [reply composer.json]: Not a bad habit at all, but that update would be unrelated to this PR, so it ought to be a separate PR in that case (which would also update the readme + the travis script to match the changed dependency requirements)

**Comment by jrfnl** [comment]
@dingo-d Ok, had yet another critical look. You're _so_ close now.
There were some very minor tidying up things, not worth the up-and-back, so I've just added a new commit to the PR to sort that out.

Now the only thing left is the unit tests.

You have five unit test files - none of which AFAICS will match the regex. so the regex part is untested.
The `attachment.inc` file says inline `OK, template file.`, but is listed in the `getErrorList()` method in the group of files to expect an error, while at the same time, it has a comment within that method as if it shouldn't.
For `attachment.inc` to match the regex would need to be updated in the `FileNameSniff`.

A filename like `page-recent-news.php` would already match and could make for a valid test case.

The other three files in the main directory all test the same thing: the fall through, so only one of those would suffice.

Another test case which could be added would be to have a `/ThemeTemplatesException/template-parts/subset/my-file-name.inc` file - this would test non-template test file names in subdirectories of one of the allowed folders are correctly excluded (which AFAICS they currently are).

If you could have a think about that, I'll have a look at the template hierarchy and see what has changed/been updated since the regex was created and see about updating the regex in WPCS.

**Comment by joyously** [comment]
Themes aren't supposed to use page template names like `page-anything.php`. Is the one in the test case supposed to be allowed or we're testing a different thing here?

**Comment by jrfnl** [comment]
> Themes aren't supposed to use page template names like page-anything.php.

It's an example which literally comes from the theme handbook: https://developer.wordpress.org/themes/basics/template-hierarchy/#single-page - second item in the list.

> Is the one in the test case supposed to be allowed or we're testing a different thing here?

Yes, it is supposed to be allowed.

**Comment by joyously** [comment]
Right, the theme handbook is for themes in general, and not specifically about themes in the repository.
A file like `page-xxx.php` should be recognized as a template file, but should not be in a theme in the repository.

**Comment by jrfnl** [comment]
> A file like page-xxx.php should be recognized as a template file, but should not be in a theme in the repository.

In that case, for the purposes of _this_ sniff, I'd say the sniff should be agnostic on whether or the theme would be allowed in the repository.

If it would be so desired that the sniffer would check for a list of allowed template file names and throw warnings for `page-xxx.php` type template files, that should be in another (new) sniff.
If you think a sniff like that should be added, let's see if there's an open issue for it and if not, open one.

**Comment by jrfnl** [comment]
> I'll have a look at the template hierarchy and see what has changed/been updated since the regex was created and see about updating the regex in WPCS.

Ok, looking more closely at the regex in WPCS, it _only_ looks for the theme template files which are allowed to break the file name rule of "no underscore", i.e. `page-post_type.php` and the likes, so adjusting that regex is not really an option.

In other words, we need to solve it here.

**Comment by dingo-d** [comment]
@jrfnl I'm happy if you can take this sniff over, it will go a lot quickly than going back and forth with the comments 🙂 

I'm puzzled about the comment about the `page-xxx.php` not being allowed in the theme, and I am failing to see the rationale behind it...

**Comment by ernilambar** [comment]
> I'm puzzled about the comment about the `page-xxx.php` not being allowed in the theme, and I am failing to see the rationale behind it...

See https://developer.wordpress.org/themes/template-files-section/page-template-files/#creating-custom-page-templates-for-global-use

**Comment by dingo-d** [comment]
This looks good to me, feel free to merge it 👍 Thanks for the help with this, I do appreciate it.

**Comment by jrfnl** [comment]
FYI: As discussed with @dingo-d on Slack, I have rebased the PR and squashed all the commits into one commit and added a commit message explaining the change in more detail.

Once the build passes for this new commit, this PR will be merged.
